### ISSUE TITLE
Improvements to tree visualization.

### ISSUE THREAD

**Issue by jtibshirani** [grf-labs/grf#298]
We added basic support for plotting trees in #296. There are a few improvements we'd like to make to the visualization:
 - [x] If the supplied data frame has column names, we should use these names when describing interior nodes. Currently, we mention the feature by its index (e.g. split_variable 5 <= 0.67). #305
- [x] We'd like to display relevant sufficient statistics in leaves, like average outcome and average treatment in the case of causal forests. This will help in debugging and in getting a better sense for the algorithm. Note that these statistics are not currently returned as part of the R function `get_tree`, so some work will be required to make them available. #460

Relates to #295, where we discuss returning relevant statistics during prediction.

**Comment by ras44** [comment]
@halflearned @jtibshirani this feature would be useful for me.  I would be happy to contribute an initial PR if it would be helpful.

**Comment by jtibshirani** [comment]
@ras44 that's great to hear. The work that's currently left is add support for useful statistics in leaves, such as the average outcome and average treatment. Let us know if you have any questions as you look into a PR.

**Comment by ras44** [comment]
@jtibshirani sounds good.  Here's my thinking after a brief look at it:

- to calculate summary statistics we would need to pass the forest object from which the tree was derived
  - this would provide the source data as well as the class of forest so that the plots can be tailored to the type of forest

- two options for plotting:
  - we could pass the forest to plot.grf_tree() as an additional argument
  - we could have individual class-specific functions like plot.quantile_forest or plot.grf that take the forest as well as a vector of tree indices to plot one or more trees in a way that is tailored to the class (while still using the modular functionality of a lower-level tree-traversal algorithm so we don't repeat the same code, of course)

Any thoughts or preferences?

**Comment by ras44** [comment]
@jtibshirani revisiting this, it looks like the first option ("... pass the forest to plot.grf_tree...") wouldn't work, as we would need to pass the index of the tree as well... as in, `plot.grf_tree(tree, forest, index)` which seems very cumbersome.

I'll develop the `plot.grf` idea using indices of the trees to plot, ie. `plot.grf(q.forest, c(1))` plots the first tree in the q.forest.  A separate generic calculates class-specific `tree_summary_stats` which can then be merged with the tree for visualization.

**Comment by jtibshirani** [comment]
> If it's ok with you, I'll develop the `plot.grf` idea using indices of the trees to plot, ie. `plot.grf(q.forest, c(1))` plots the first tree in the q.forest.

This works for me, just a couple comments to make sure the idea is clear:
- We can move the plotting functionality over from `plot.grf_tree` to this new method `plot.grf(forest, index)`.
- Perhaps we could just accept one index for now, instead of a list? I noticed you wrote `c(1)` as the second argument to `plot.grf`, w hich suggests multiple indices could be provided.

> A separate generic calculates class-specific `tree_summary_stats` which can then be merged with the tree for visualization.

Although generic functions are often the proper 'object oriented' solution, it would be nice to keep the design as simple as possible in this first iteration. When deciding what summary statistics to compute, we could simply check the class of the forest (through `class(forest)`) instead of introducing a new generic function.

**Comment by ras44** [comment]
@jtibshirani I've pushed a new branch with a first version that includes an example of calculating stats for leaves.  I had already started the generic function version, but I'd be happy to switch it to the `class(forest)` logic if the generics seem too cumbersome.  I did keep things simple and just use one index.

An obvious next step would be to decide which stats to include for specific forest classes.  An additional possible feature would be to allow the user to pass custom leaf-calculation functions in.  Looking forward to your thoughts!
